internal/apimeta/mapper.go

```diff
@@ -0,0 +1,81 @@
+package apimeta
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+// ResourceMapper allows callers to map resources to kinds.
+type ResourceMapper interface {
+	// NamespacedKindFor returns the priority mapping GVK for a namespaced GR.
+	// If the GR doesn't exist, or isn't namespaced, returns an error.
+	NamespacedKindFor(gr schema.GroupResource) (schema.GroupVersionKind, error)
+}
+
+// NotNamespacedError is returned if the mapper can find the resource, but it is not namespaced.
+type NotNamespacedError struct {
+	GroupResource schema.GroupResource
+}
+
+func (e *NotNamespacedError) Error() string {
+	return fmt.Sprintf("resource %v is not namespaced", e.GroupResource)
+}
+
+func IsNotNamespacedError(err error) bool {
+	if err == nil {
+		return false
+	}
+	switch err.(type) {
+	case *NotNamespacedError:
+		return true
+	default:
+		return false
+	}
+}
+
+func NewAPIMapper(c *rest.Config) (ResourceMapper, error) {
+	restMapper, err := apiutil.NewDynamicRESTMapper(c)
+	if err != nil {
+		return nil, err
+	}
+	return &APIMapper{
+		RestMapper: restMapper,
+	}, nil
+
+}
+
+// APIMapper allows callers to discover apiserver metadata at runtime
+type APIMapper struct {
+	RestMapper meta.RESTMapper
+}
+
+// NamespacedKindFor implements ResourceMapper.NamespacedKindFor by using a DynamicRESTMapper
+// to discover resource types at runtime.
+func (r *APIMapper) NamespacedKindFor(gr schema.GroupResource) (schema.GroupVersionKind, error) {
```

Comment:
I don't really see the need to have `Namespaced` in the name? You can just comment it. Also this doesn't _just_ return the kind, so maybe `GetGVK` or (if you think it's important to keep) `GetNamespacedGVK`? (GVK is a fairly well-known acronym for GroupVersionKind).

Reply:
The term `Namespaced` was introduced to make the caller aware that the func will error out if you ask for a non-namespaced GVK-mapping.

I am personally not a big fan of the `Get`/`Set` func-names - unless they are **only**  getting/setting. So I tried to build on the IMO meaningful function names in `RestMapper`.

Reply:
I'd kind of like this function to only return the `Kind` if it keeps the current name, or include `GVK` in its name if it continues to return a GVK. Beyond that, up to you.

Reply:
It returns a [kind](https://kubernetes.io/docs/reference/using-api/api-concepts/#standard-api-terminology), fully qualified as a GVK. 😉  I do not see the need to use the GVK term when naming a function. The return type is quite clear. I prefer including "namespaced" in the name, but I'll change this func-name to whatever you wish - to get this merged. 😃